hazelcast/src/test/java/com/hazelcast/map/impl/eviction/ExpirationManagerTest.java

```diff
@@ -300,43 +314,90 @@ public void run() throws Exception {
+
+        IMap map = liteMember.getMap("test");
+        map.put(1, 1, 3, TimeUnit.SECONDS);
+
+        assertTrueAllTheTime(new AssertTask() {
+            @Override
+            public void run() {
+                assertFalse("There should be zero ClearExpiredRecordsTask",
+                        hasClearExpiredRecordsTaskStarted(liteMember));
+            }
+        }, 3);
     }
 
     @Test
-    public void restarts_running_backgroundClearTask_when_lifecycleState_turns_to_MERGED() {
+    public void clearExpiredRecordsTask_should_be_started_when_mapConfig_ttl_expiry() {
+        String mapName = "test";
+
         Config config = new Config();
         config.setProperty(PROP_TASK_PERIOD_SECONDS, "1");
+        config.getMapConfig(mapName).setTimeToLiveSeconds(2);
         HazelcastInstance node = createHazelcastInstance(config);
 
-        final AtomicInteger expirationCounter = new AtomicInteger();
+        IMap map = node.getMap(mapName);
+        map.put(1, 1);
 
-        IMap map = node.getMap("test");
-        map.addEntryListener(new EntryExpiredListener() {
-            @Override
-            public void entryExpired(EntryEvent event) {
-                expirationCounter.incrementAndGet();
-            }
-        }, true);
+        assertTrue("There should be one ClearExpiredRecordsTask",
+                hasClearExpiredRecordsTaskStarted(node));
+    }
 
-        map.put(1, 1, 3, TimeUnit.SECONDS);
+    @Test
+    public void clearExpiredRecordsTask_should_be_started_when_mapConfig_has_idle_expiry() {
+        String mapName = "test";
 
-        ((LifecycleServiceImpl) node.getLifecycleService()).fireLifecycleEvent(MERGING);
-        ((LifecycleServiceImpl) node.getLifecycleService()).fireLifecycleEvent(MERGED);
+        Config config = new Config();
+        config.setProperty(PROP_TASK_PERIOD_SECONDS, "1");
+        config.getMapConfig(mapName).setMaxIdleSeconds(2);
+        HazelcastInstance node = createHazelcastInstance(config);
 
-        assertTrueEventually(new AssertTask() {
-            @Override
-            public void run() throws Exception {
-                int expirationCount = expirationCounter.get();
-                assertEquals(format("Expecting 1 expiration but found:%d", expirationCount), 1, expirationCount);
+        IMap map = node.getMap(mapName);
+        map.put(1, 1);
+
+        assertTrue("There should be one ClearExpiredRecordsTask",
+                hasClearExpiredRecordsTaskStarted(node));
+    }
+
+    private boolean hasClearExpiredRecordsTaskStarted(HazelcastInstance node) {
+        ExecutionServiceImpl executionService = (ExecutionServiceImpl) getNodeEngineImpl(node).getExecutionService();
+        LoggingScheduledExecutor scheduledExecutorService = executionService.getScheduledExecutorService();
+        BlockingQueue<Runnable> queue = scheduledExecutorService.getQueue();
+        for (Runnable runnable : queue) {
+            if (runnable.toString().contains("ClearExpiredRecordsTask")) {
```

Comment:
This looks a bit brittle (if we would rename the class, this test would break silently).

Can't we use the `getExpirationManager()` from the other test and check the scheduled state?

Reply:
fixed